ocaml/quicktest/qt.ml

```diff
@@ -0,0 +1,217 @@
+end
+
+module Time = struct
+  type t = float
+
+  let now () = Unix.gettimeofday ()
+
+  let of_field = Xapi_stdext_date.Date.to_float
+
+  let pp t = Xapi_stdext_date.Date.of_float t |> Xapi_stdext_date.Date.to_string
+
+  let check t ~after ~before =
+    Alcotest.(check bool)
+      (Printf.sprintf "Time %s should be between %s and %s (+-1s)" (pp t) (pp before) (pp after))
+      true
+      (t > (after -. 1.0) && t < (before +. 1.0))
+end
+
+module VM = struct
+  module Template = struct
+    exception Unable_to_find_suitable_vm_template
+
+    let other = "Other install media"
+
+    let find rpc session_id startswith =
+      let vms = Client.Client.VM.get_all rpc session_id in
+      match List.filter (fun self ->
+          (Xapi_stdext_std.Xstringext.String.startswith startswith (Client.Client.VM.get_name_label rpc session_id self))
+          && (Client.Client.VM.get_is_a_template rpc session_id self)
+        ) vms with
+      | [] -> raise Unable_to_find_suitable_vm_template
+      | x :: _ ->
+        Printf.printf "Choosing template with name: %s\n" (Client.Client.VM.get_name_label rpc session_id x);
+        x
+  end
+
+  let install rpc session_id template name =
+    let newvm_uuid = cli_cmd [ "vm-install"; "template-uuid=" ^ template; "new-name-label=" ^ name ] in
+    Client.Client.VM.get_by_uuid rpc session_id newvm_uuid
+
+  let dom0_of_host rpc session_id host =
+    Client.Client.Host.get_control_domain rpc session_id host
+
+  let get_dom0 rpc session_id =
+    Xapi_inventory.inventory_filename := "/etc/xensource-inventory";
+    let uuid = Xapi_inventory.lookup Xapi_inventory._control_domain_uuid in
+    Client.Client.VM.get_by_uuid ~rpc ~session_id ~uuid
+end
+
+module VDI = struct
+
+  (* This naming is used to identify VDIs to destroy later on *)
+  let test_vdi_name_label = "quicktest-vdi_16cd61e7-3f42-4ea7-a1d8-5e9d3f64f25d"
+  let test_vdi_name_description = "VDI for storage quicktest"
+
+  let make rpc session_id ?(virtual_size=4194304L) sR =
+    Client.Client.VDI.create
+      ~sR
+      ~session_id
+      ~rpc
+      ~name_label:test_vdi_name_label
+      ~name_description:test_vdi_name_description
+      ~_type:`user
+      ~sharable:false
+      ~read_only:false
+      ~virtual_size
+      ~xenstore_data:[]
+      ~other_config:[]
+      ~tags:[]
+      ~sm_config:[]
+
+  let with_destroyed rpc session_id self f =
+    Xapi_stdext_pervasives.Pervasiveext.finally
+      f
+      (fun () -> Client.Client.VDI.destroy ~rpc ~session_id ~self)
+
+  let with_new rpc session_id ?(virtual_size=4194304L) sr f =
+    let self = make rpc session_id ~virtual_size sr in
+    with_destroyed rpc session_id self (fun () -> f self)
+
+  let with_any rpc session_id sr_info f =
+    if List.mem `vdi_create sr_info.allowed_operations then begin
+      with_new rpc session_id sr_info.sr f
+    end else begin
+      let self = Client.Client.SR.get_VDIs ~rpc ~session_id ~self:sr_info.sr |> List.hd in
+      f self
+    end
```

Comment:
Don't think we need `begin ... end` anywhere in here?

Reply:
True, I left it there to make the code clearer because we have more than one line here in the else case.